x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/AutodetectProcessManager.java

```diff
@@ -500,6 +511,17 @@ protected void doRun() {
                             closeHandler.accept(null);
                             return;
                         }
+                        if (resetInProgress) {
+                            closeHandler.accept(
+                                new ElasticsearchStatusException(
+                                    "cancelling model upgrade for snapshot [{}] for job [{}] due to feature reset",
+                                    RestStatus.NOT_FOUND,
+                                    jobId,
+                                    snapshotId
+                                )
+                            );
+                            return;
+                        }
```

Comment:
I think this should be more like the `nodeDying` case above, where it returns `null` to the close handler.

Returning an exception here will cause a warning to be logged in the calling code.  We should probably go even further than the `nodeDying` case, and not log it at all (or log at trace level), because if the reset had happened just a second later then all its work would have been silently wiped out by the feature reset.  Also, returning an exception here causes a notification message to be indexed.  Since that's async that might be the next subtle race condition, or even if we're protecting against notification logging messing up feature reset it's still pointless work that's immediately undone.